### PR TITLE
[op-conductor] state transition bug fix

### PR DESCRIPTION
**Pull request by 0x00101010**
**Description**

This PR fixes a state transition bug described in https://github.com/ethereum-optimism/protocol-quest/issues/114 as `edge case scenario 1`.

***edge case scenario 1***
Currently our state transition logic will directly try to transfer leadership to another node if current node is a leader, is not healthy and not sequencing. However, during edge cases situation where:

active sequencer is down
new leader was elected after relatively long time (flaky in CI, could happen in real world as well)
all the followers became unhealthy before becoming leader
In the above situation, we'll enter into an infinite leadership transfer loop with no node being able to start sequencing.

***Changes***

1. reworked core control loop logic to handle the edge case, introduced the concept of previous state (because not only do we need the current state to decide which action to take, we'll also need previous state as well, see code comment for more details)
2. reworked test synchronization logic, instead of instrumenting just on the action step, instrumenting around the entire action loop so that we can have more granular control
3. Added a new test case to cover the edge case
4. Added a simple script to help run conductor tests to detect race condition situations.
5. Also fixed a monitor bug

**Tests**

All tests are updated to cover the rework and the new edge case.

**Additional context**

This is discovered during e2e testing

**Metadata**

- https://github.com/ethereum-optimism/protocol-quest/issues/114

### CODE REVIEW

File: op-conductor/run_test_100times.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+set -e
+
+for i in {1..100}; do
+  echo "======================="
+  echo "Running iteration $i"
+  gotestsum -- -run 'TestControlLoop' ./... --count=1 --timeout=5s -race
+  if [ $? -ne 0 ]; then
+    echo "Test failed"
+  fi
+done
```
[the remark this answers is not in the document]
Comment (0x00101010) [reply op-conductor/run_test_100times.sh]: done

File: op-conductor/conductor/service.go [hunk 2]
```diff
@@ -68,7 +68,7 @@ func NewOpConductor(
 		cons:         cons,
 		hmon:         hmon,
 	}
-	oc.actionFn = oc.action
+	oc.loopActionFn = oc.loopAction
```
[the remark this answers is not in the document]
Comment (0x00101010) [reply op-conductor/conductor/service.go:71]: This is intentional to avoid anybody accidentally trying to configure loop action in code.

The configuration will only be done in internal tests

File: op-conductor/conductor/service.go [hunk 3]
```diff
@@ -252,10 +253,11 @@ type OpConductor struct {
 	seqActive atomic.Bool
 	healthy   atomic.Bool
 	hcerr     error // error from health check
+	prevState *state
 
 	healthUpdateCh <-chan error
 	leaderUpdateCh <-chan bool
-	actionFn       func() // actionFn defines the action to be executed to bring the sequencer to the desired state.
+	loopActionFn   func() // loopActionFn defines the logic to be executed inside control loop.
```
[the remark this answers is not in the document]
Comment (0x00101010) [reply op-conductor/conductor/service.go:260]: There is?

File: op-conductor/conductor/service.go [hunk 4]
```diff
@@ -536,17 +591,25 @@ func (oc *OpConductor) action() {
...
 	}
 
+	oc.log.Debug("exiting action with status and error", "status", status, "err", err)
 	if err != nil {
 		oc.log.Error("failed to execute step, queueing another one to retry", "err", err)
 		// randomly sleep for 0-200ms to avoid excessive retry
 		time.Sleep(time.Duration(rand.Intn(200)) * time.Millisecond)
 		oc.queueAction()
+		return
+	}
```
[the remark this answers is not in the document]
Comment (0x00101010) [reply op-conductor/conductor/service.go:601]: This is outside the scope of this PR.

File: op-conductor/conductor/service_test.go [hunk 5]
```diff
@@ -79,21 +79,22 @@ func mockConfig(t *testing.T) Config {
...
 	version string
 	ctrl    *clientmocks.SequencerControl
 	cons    *consensusmocks.Consensus
 	hmon    *healthmocks.HealthMonitor
 
-	next chan struct{}
-	wg   sync.WaitGroup
+	syncEnabled bool
+	next        chan struct{}
+	wg          sync.WaitGroup
```
[the remark this answers is not in the document]
Comment (0x00101010) [reply op-conductor/conductor/service_test.go]: done

File: op-conductor/conductor/service_test.go [hunk 6]
```diff
@@ -115,57 +116,75 @@ func (s *OpConductorTestSuite) SetupTest() {
...
 }
 
-func (s *OpConductorTestSuite) updateLeaderStatusAndExecuteAction(ch chan bool, status bool) {
+func updateStatusAndExecuteAction[T any](s *OpConductorTestSuite, ch chan T, status T) {
 	fn := func() {
 		ch <- status
 	}
-	s.execute(fn)
+	s.execute(fn) // this executes status update
+	s.executeAction()
```
[the remark this answers is not in the document]
Comment (0x00101010) [reply op-conductor/conductor/service_test.go:179]: Could be helpful for test execution, however, sometimes we might need to wait for certain period of time, so elect not to add this one for now.

**Review by zhwrd — APPROVED** [review]

File: op-conductor/health/monitor_test.go [hunk 7]
```diff
@@ -26,130 +26,192 @@ const (
...
+			Number: safeNum,
+		},
+	}
 }
 
 func TestHealthMonitor(t *testing.T) {
 	suite.Run(t, new(HealthMonitorTestSuite))
 }
+
+type timeProvider struct {
```
Comment (mslipper) [inline op-conductor/health/monitor_test.go:209]: [nit] I believe there's already a fake clock utility in `op-service` you can use

**Review by mslipper — APPROVED** [review]

File: op-conductor/health/monitor_test.go (hunk 7 above)
Comment (0x00101010) [reply op-conductor/health/monitor_test.go:209]: Thanks, just checked `DeterministicClock`, looks good. Can rework in a later PR

**Review by trianglesphere — APPROVED** [review]
